mono/metadata/class.c

```diff
@@ -2491,33 +2480,40 @@ mono_class_get_method_by_index (MonoClass *klass, int index)
 	}
 }	
 
-/*
+/**
  * mono_class_get_inflated_method:
+ * \param klass an inflated class
+ * \param method a method of \p klass's generic definition
+ * \param error set on error
  *
- *   Given an inflated class CLASS and a method METHOD which should be a method of
- * CLASS's generic definition, return the inflated method corresponding to METHOD.
+ * Given an inflated class \p klass and a method \p method which should be a
+ * method of \p klass's generic definition, return the inflated method
+ * corresponding to \p method.
+ *
+ * On failure sets \p error and returns NULL.
  */
 MonoMethod*
-mono_class_get_inflated_method (MonoClass *klass, MonoMethod *method)
+mono_class_get_inflated_method (MonoClass *klass, MonoMethod *method, MonoError *error)
 {
 	MonoClass *gklass = mono_class_get_generic_class (klass)->container_class;
 	int i, mcount;
 
 	g_assert (method->klass == gklass);
+	error_init (error);
 
 	mono_class_setup_methods (gklass);
-	g_assert (!mono_class_has_failure (gklass)); /*FIXME do proper error handling*/
+	if (mono_class_has_failure (gklass)) {
+		mono_error_set_for_class_failure (error, gklass);
+		return NULL;
+	}
 
 	mcount = mono_class_get_method_count (gklass);
 	for (i = 0; i < mcount; ++i) {
 		if (gklass->methods [i] == method) {
 			if (klass->methods) {
 				return klass->methods [i];
```

Comment:
I believe that klass->methods can have null entries due to loading errors even though we didn't fail the class.

We should either check/fail or assert around that.

Reply:
I made a pass over `class.c` and I don't see any obvious ways that `klass->methods[i]` could ever be NULL.

Also I see a lot of places where we do something like `klass->methods[i]->....` guarded only by a check that `klass->methods != NULL`.

In any case, I'll assert that `mono_class_get_inflated_method` only returns NULL when `error` is set.